File: ios/MapPolygonManager.swift
import GoogleMaps

private final class PolygonState {
  var current: RNPolygon
  var polygon: GMSPolygon?

  init(current: RNPolygon) {
    self.current = current
  }
}

final class MapPolygonManager {
  private let builder: MapPolygonBuilder
  private weak var mapView: GMSMapView?
  private var states: [String: PolygonState] = [:]
  private var destroyed = false

  init(builder: MapPolygonBuilder) {
    self.builder = builder
  }

  func attachMap(_ mapView: GMSMapView) {
    onMain {
      guard !self.destroyed else { return }
      self.mapView = mapView
      self.states.values.filter { $0.polygon == nil }.forEach { self.addToMap($0) }
    }
  }

  func add(_ polygon: RNPolygon) {
    onMain {
      guard !self.destroyed else { return }
      self.states.removeValue(forKey: polygon.id).map { self.removeFromMap($0) }
      let state = PolygonState(current: polygon)
      self.states[polygon.id] = state
      if self.mapView != nil {
        self.addToMap(state)
      }
    }
  }

  func update(_ next: RNPolygon) {
    onMain {
      guard !self.destroyed else { return }
      guard let state = self.states[next.id] else { return }
      let prev = state.current
      if prev.polygonEquals(next) { return }
      state.current = next
      state.polygon.map { self.builder.update(prev, next, $0) }
    }
  }

  func remove(id: String) {
    onMain {
      self.states.removeValue(forKey: id).map { self.removeFromMap($0) }
    }
  }

  func destroy() {
    onMain {
      self.destroyed = true
      self.states.values.forEach { self.removeFromMap($0) }
      self.states.removeAll()
      self.mapView = nil
    }
  }

  private func addToMap(_ state: PolygonState) {
    guard let mapView = self.mapView else { return }
    let polygon = self.builder.build(state.current)
    polygon.tagData = PolygonTag(id: state.current.id)
    polygon.map = mapView
    state.polygon = polygon
  }

  private func removeFromMap(_ state: PolygonState) {
    state.polygon?.map = nil
  }
}

File: android/src/main/java/com/rngooglemapsplus/MapPolygonManager.kt
package com.rngooglemapsplus

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Polygon
import com.rngooglemapsplus.extensions.polygonEquals

private class PolygonState(
  var current: RNPolygon,
) {
  var polygon: Polygon? = null
}

class MapPolygonManager(
  private val builder: MapPolygonBuilder,
) {
  private var map: GoogleMap? = null
  private val states = mutableMapOf<String, PolygonState>()
  private var destroyed = false

  fun attachMap(map: GoogleMap) =
    onUi {
      if (destroyed) return@onUi
      this.map = map
      states.values.filter { it.polygon == null }.forEach { addToMap(it) }
    }

  fun add(polygon: RNPolygon) =
    onUi {
      if (destroyed) return@onUi
      remove(polygon.id)
      val state = PolygonState(polygon)
      states[polygon.id] = state
      if (map != null) addToMap(state)
    }

  fun update(next: RNPolygon) =
    onUi {
      if (destroyed) return@onUi
      val state = states[next.id] ?: return@onUi
      val prev = state.current
      if (prev.polygonEquals(next)) return@onUi
      state.current = next
      state.polygon?.let { builder.update(prev, next, it) }
    }

  fun remove(id: String) =
    onUi {
      states.remove(id)?.let { removeFromMap(it) }
    }

  fun destroy() =
    onUi {
      destroyed = true
      states.values.forEach { removeFromMap(it) }
      states.clear()
      map = null
    }

  private fun addToMap(state: PolygonState) {
    state.polygon =
      map?.addPolygon(builder.build(state.current))?.apply {
        tag = PolygonTag(id = state.current.id)
      }
  }

  private fun removeFromMap(state: PolygonState) {
    state.polygon?.remove()
  }
}
